Clarification on custody sampling (#4251)
ethereum/consensus-specs: specs/fulu/das-core.md

| Name                       | Value | Description                                                                       |
| -------------------------- | ----- | --------------------------------------------------------------------------------- |

> | `SAMPLES_PER_SLOT`         | `8`   | Minimum number of custody groups an honest node samples from                      |

Reviewer: ```suggestion
| `SAMPLES_PER_SLOT`         | `8`   | Minimum number of samples for an honest node                       |
```

I would consider changing `SAMPLES_PER_SLOT` to not specify the units, so that it can be used by both custody sampling (sample = custody group) and the separate peer sampling spec (sample = column)

## Custody sampling

> At each slot, a node advertising `custody_group_count` downloads a minimum of `sampling_size = max(SAMPLES_PER_SLOT, custody_group_count * columns_per_group)` total columns, where `columns_per_group = NUMBER_OF_COLUMNS // NUMBER_OF_CUSTODY_GROUPS`. The corresponding set of columns is selected by `groups = get_custody_groups(node_id, sampling_size // columns_per_group)` and `compute_columns_for_custody_group(group) for group in groups`, so that in particular the subset of columns to custody is consistent with the output of `get_custody_groups(node_id, custody_group_count)`. Sampling is considered successful if the node manages to retrieve all selected columns.

Reviewer 2: Hmm this isn't quite right. We need to round up. For example, if `sampling_size` is 7 and `columns_per_group` is 2 then only `7 // 2 = 1` groups would be selected.

Reviewer: So replace it with `(sampling_size + columns_per_group - 1) // columns_per_group`?

Reviewer: Even with rounding up, this is not correct. If `columns_per_group = 2` and everything else is as today, `sampling_size` would be 8. In that case, we don't just want to get any 8 columns, we want 8 randomly selected columns, so it's not okay to randomly choose just 4 custody groups. If we can't get individual columns and have to download whole custody groups (in practice because we get columns through subnets), then we'd need to get 8 custody groups, and 16 columns (we only care about 8 randomly selected columns, but we have no way to just get 8). In this case, `get_custody_groups(node_id, sampling_size)` would be right.

However, if `columns_per_custody_group = 4`, `sampling_size` would be 16, which would be wrong regardless of whether we interpret it as columns or custody groups. If we interpret it as columns, corresponding to `sampling_size // columns_per_group = 4` custody groups, then it's again too few, because we need 8 randomly selected columns. If we interpret it as custody groups, it's too many, because we only need 8 custody groups, not 16. 

```suggestion
At each slot, a node advertising `custody_group_count` downloads a minimum of `sampling_size = max(SAMPLES_PER_SLOT, custody_group_count) * columns_per_group` columns, where `columns_per_group = NUMBER_OF_COLUMNS // NUMBER_OF_CUSTODY_GROUPS`.  The columns are selected by `groups = get_custody_groups(node_id, sampling_size)`, and `compute_columns_for_custody_group(group) for group in groups`, so that in particular the subset of columns to custody is consistent with the output of `get_custody_groups(node_id, custody_group_count)`. Sampling is considered successful if the node manages to retrieve all selected columns.
```suggestion

This is correct in that the sampling unit is now custody groups. However, it doesn't play well with the accompanying peer sampling spec, because for peer sampling you'd want the extra sampling to have nothing to do with custody groups, and be measured in columns (as is the case there now). On the other hand, using columns as the sampling unit here is problematic because it really needs to be columns that are *independently* randomly selected. 

Given that the peer sampling spec anyway does things differently, with the sample choice being done with local randomness that changes every slot rather than through `node_id` (meaning, the sample choice from here doesn't generalize to peer sampling, and we already don't have a nice "these are always the samples you want, then it's up to you how you get them"), my preference would be to go with the simplest choice here, which I think is using custody groups as the sampling unit (the suggestion above).

Author: > Hmm this isn't quite right. We need to round up. For example, if `sampling_size` is 7 and `columns_per_group` is 2 then only `7 // 2 = 3` groups (6 columns, fewer than the required 7 columns) would be selected.

I assumed `columns_per_group` divides `sampling_size`.

Author: @fradamt If you think the sampling unit should be custody groups, your change makes sense then.

Author: @fradamt If you would like to apply your change, the unit of `SAMPLES_PER_SLOT` should be custody groups as well, right? does that make sense?

Reviewer: Right. So:

```suggestion
At each slot, a node advertising `custody_group_count` downloads a minimum of `sampling_size = max(SAMPLES_PER_SLOT, custody_group_count)` custody groups, selected by `groups = get_custody_groups(node_id, sampling_size)`, to which correspond the columns `compute_columns_for_custody_group(group) for group in groups`.  The custody groups (and corresponding columns) to custody, selected by `get_custody_groups(node_id, custody_group_count)`, are then a subset of those to sample. Sampling is considered successful if the node manages to retrieve all selected columns.
```

Author: Yeah, that makes sense now. I will update the definition of `SAMPLES_PER_SLOT` correspondingly.

Author: @fradamt https://github.com/ethereum/consensus-specs/pull/4251/commits/3096c4eca14f80c48a813593f044fa7eae0f378d Does this look good?
